mlir/lib/Catalyst/Transforms/DetensorizeFunctionBoundaryPass.cpp

```diff
@@ -0,0 +1,309 @@
+                // Insert a tensor extract operation to bridge scalar tensor to scalar for return
+                auto extractOp =
+                    rewriter.create<tensor::ExtractOp>(returnOp.getLoc(), operand, ValueRange{});
+                newOperands.push_back(extractOp.getResult());
+            }
+            else {
+                newOperands.push_back(operand);
+            }
+        }
+
+        rewriter.replaceOpWithNewOp<func::ReturnOp>(returnOp, newOperands);
+        return success();
+    }
+};
+
+struct DetensorizeCallPattern : public OpRewritePattern<func::CallOp> {
+    using OpRewritePattern<func::CallOp>::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(func::CallOp callOp, PatternRewriter &rewriter) const override
+    {
+        // Only update the call op if it differs from the function signature
+        auto module = callOp->getParentOfType<ModuleOp>();
+        auto funcOp = module.lookupSymbol<func::FuncOp>(callOp.getCallee());
+        if (!funcOp) {
+            return failure();
+        }
+
+        FunctionType funcType = funcOp.getFunctionType();
+        if (callOp.getCalleeType() == funcType) {
+            return failure();
+        }
+
+        rewriter.setInsertionPoint(callOp);
+        SmallVector<Value> newOperands;
+        for (unsigned i = 0; i < callOp.getNumOperands(); ++i) {
+            Value operand = callOp.getOperand(i);
+            if (isScalarTensor(operand.getType()) && operand.getType() != funcType.getInput(i)) {
+                // Insert extract op if the operand is converted from tensor to scalar
+                auto extractOp =
+                    rewriter.create<tensor::ExtractOp>(callOp.getLoc(), operand, ValueRange{});
+                newOperands.push_back(extractOp.getResult());
+            }
+            else {
+                newOperands.push_back(operand);
+            }
+        }
+
+        auto newCallOp = rewriter.create<func::CallOp>(callOp.getLoc(), funcOp, newOperands);
+
+        SmallVector<Value> newResults;
+        for (unsigned i = 0; i < callOp.getNumResults(); ++i) {
+            Value oldResult = callOp.getResult(i);
+            Value newResult = newCallOp.getResult(i);
+            if (isScalarTensor(oldResult.getType()) && oldResult.getType() != newResult.getType()) {
+                // Insert FromElement op if result is converted from tensor to scalar
+                auto fromElementsOp = rewriter.create<tensor::FromElementsOp>(
+                    callOp.getLoc(), oldResult.getType(), newResult);
+                newResults.push_back(fromElementsOp.getResult());
+            }
+            else {
+                newResults.push_back(newResult);
+            }
+        }
+
+        rewriter.replaceOp(callOp, newResults);
+        return success();
+    }
+};
+
+// Remove from_elements -> extract
+struct FoldExtractFromElementsPattern : public OpRewritePattern<tensor::ExtractOp> {
```

Comment:
Out of curiosity this pattern doesn't exist upstream? Either way though probably no harm in having it here explicitly